File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private long sum(TreeNode node)
    {
        if(node == null)
        {
            return 0;
        }
        node.val +=sum(node.left) + sum(node.right);
        return node.val;
    }
    public int maxProduct(TreeNode root) {
        long totalSum = sum(root);
        Queue<TreeNode> q = new ArrayDeque<>();
        q.add(root);
        long ans =0;
        long MOD = 1000000007L;
        while(!q.isEmpty())
        {
            TreeNode node = q.poll();
            if(node == null)
            {
                continue;
            }
            long res = (totalSum - node.val)*node.val;
            ans = Math.max(ans, res);
            System.out.println(node.val);
           if (node.left != null) q.add(node.left);
           if (node.right != null) q.add(node.right);
        }
        return (int)(ans % MOD);
    }
}

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
int mod = 1000000007;
int totalSum =0;
long long ans=0;
    long long findSum(TreeNode* root)
    {
        if(!root)
        {
            return 0;
        }
        long long leftSum = findSum(root->left);
        long long rightSum = findSum(root->right);
        long long total = ((leftSum + rightSum) + root->val);
        return total;
    }
    long long solve(TreeNode* root)
    {
        if(root == nullptr)
        {
            return 0;
        }
        long long leftSum = solve(root->left);
        long long rightSum = solve(root->right);
        long long op1= ((totalSum-leftSum)*1LL*leftSum);
        long long op2 = ((totalSum-rightSum)*1LL*rightSum);
        ans=max(ans,max(op1,op2));
        long long total = ((leftSum + rightSum) + root->val);
        return total;
    }
    int maxProduct(TreeNode* root) {
    totalSum = findSum(root);
    solve(root);
    return (int)(ans%mod);   
    }
};